### ISSUE TITLE
Province/state as target

### ISSUE THREAD

**Issue by missmatsuko** [doublesecretagency/craft-smartmap#41]
Is it possible to search address with a province or state as a target? 
The docs say the target can be anything that translates into a full or partial address, but I'm not getting results when searching for a province and I'm not sure if that's expected.

e.g. searching with a target as "Alberta" or "AB".

If I call Google's geocode service directly, I do get a response with location coordinates for 'Alberta':
```
{
   "results" : [
      {
         "address_components" : [
            {
               "long_name" : "Alberta",
               "short_name" : "AB",
               "types" : [ "administrative_area_level_1", "political" ]
            },
            {
               "long_name" : "Canada",
               "short_name" : "CA",
               "types" : [ "country", "political" ]
            }
         ],
         "formatted_address" : "Alberta, Canada",
         "geometry" : {
            "bounds" : {
               "northeast" : {
                  "lat" : 60.00006209999999,
                  "lng" : -109.9998551
               },
               "southwest" : {
                  "lat" : 48.9966671,
                  "lng" : -120.0005219
               }
            },
            "location" : {
               "lat" : 53.9332706,
               "lng" : -116.5765035
            },
            "location_type" : "APPROXIMATE",
            "viewport" : {
               "northeast" : {
                  "lat" : 60.00006209999999,
                  "lng" : -109.9998551
               },
               "southwest" : {
                  "lat" : 48.9966671,
                  "lng" : -120.0005219
               }
            }
         },
         "place_id" : "ChIJtRkkqIKyCVMRno6bQJpHqbA",
         "types" : [ "administrative_area_level_1", "political" ]
      }
   ],
   "status" : "OK"
}
```

**Comment by lindseydiloreto** [comment]
Very strange, that _should_ work. I'll do a little investigating, and get back to you.

**Comment by cakleimeier** [comment]
Has there been any progress or additional information on this issue? I'm running into a similar issue, where searching for US State (by name or abbreviation) only returns results if the address in that state has no street address or zip code information.

**Comment by ryanmasuga** [comment]
Please let us know what you figure out, @lindseydiloreto. We'd like to deploy a map update for a client but it's sort of dependent on being able to search a state as @cakleimeier mentioned. Thanks!

**Comment by lindseydiloreto** [comment]
I'll take a look into this today, and report back. 👍

**Comment by lindseydiloreto** [comment]
Quick question for any/all of you... When performing the search, what `range` are you using? If you bump the `range` up to something huge (ie: 100000), does it then return results?

What I suspect may be happening is that your state/province search is using the **center** of that state/province to conduct the proximity search. So basically, your query is being translated as "closest to the center of Alberta". And if your `range` isn't large enough, anything too far away from the center will simply not appear in the results.

If that's the case, you might actually need to switch up the internal logic to [filter by a subfield value...](https://www.doublesecretagency.com/plugins/smart-map/docs/filtering-entries-by-subfield-value)

Internally, your code would try to identify if the person has entered a specific address, or if they've entered a broad state/province. Depending on which type of search they are conducting, you'd then either (A) do a normal proximity search, or (B) filter by the state subfield value.

There's a chance that I'm mistaken about this, and that something larger is going on, but at this point I'm about 70% confident that we're headed in this direction.

Let me know your thoughts!

**Comment by missmatsuko** [comment]
I think that is what's happening on my site. 

Do you have any suggestions for how to determine if a user's entering a specific address or not?

Could the plugin be updated to look for results within bounds instead of using central coordinates when the `address_components` isn't very granular (like if it only has a country and administrative area)?

**Comment by lindseydiloreto** [comment]
That's interesting... maybe? It would still ping Google (obviously), but that may be a straightforward way to detect whether the address is specific enough.

Without updating the plugin, I'd recommend comparing the user's submission against a pre-set list of states/provinces...

```twig
{% set states = {
    'alabama': 'al',
    'alaska': 'ak',
    ...
    'wyoming': 'wy'
} %}

{% set target = craft.request.getParam('near') %}
{% set target = target | lower | trim %}

{% if states[target] is defined %}
    {# Searched for "Wyoming" #}
    {% set filter = {state: states[target]} %}
{% elseif target in states %}
    {# Searched for "WY" #}
    {% set filter = {state: target} %}
{% else %}
    {# Not a state #}
    {% set filter = false %}
{% endif %}

{% if filter %}
    {% set params = {
        filter: filter
    } %}
{% else %}
    {% set params = {
        target: target,
        range: 100
    } %}
{% endif %}

{% set entries = craft.entries.addressField(params).order('distance').all() %}
```

^ Pseudo-code, you may need to adjust it a bit for your needs.

Hope that helps!

**Comment by missmatsuko** [comment]
I guess that will work for me, but it will be a long list of provinces/states for anyone working on an international map (even more if multilingual).

**Comment by lindseydiloreto** [comment]
Yeah, I agree that it leads to a lot of extra code. I'll take a look into your suggestion (which is a pretty solid idea).

**Comment by cakleimeier** [comment]
@lindseydiloreto The "range" was what was causing the issue for me. Thanks!

**Comment by lindseydiloreto** [comment]
@cakleimeier Ok great, no problem! 🙂

**Comment by missmatsuko** [comment]
I'm going to try and do something in Twig similar to what you posted, but I just wanted to note that it's looking to be more complicated to do for Canada, especially for a bilingual site.

PEI is possibly one of the worst cases, since it has at least 7 variations:
- PEI
- PE
- Prince Edward Island
- Île-du-Prince-Édouard
- L’Île-du-Prince-Édouard
- Î.-P.-É.
- l’Î.-P.-É.
- All of the above ignoring accent marks and punctuation (period, hyphens)

**Comment by lindseydiloreto** [comment]
Yikes.

One other option that I've seen people do in the past... Include a dropdown menu of states/provinces next to your search bar. So if someone just selects something from the dropdown menu, and doesn't type anything into the search field, you already know to filter it instead of doing a lookup.

**Comment by missmatsuko** [comment]
Yeah, we actually already have a province filter but they want the users to be able to type it in, too.

**Comment by missmatsuko** [comment]
@tremby helped me make a patch for the plugin so when the target returns a geocoding response whose `address_components` contain `administative_area_level_1` but not `locality` (i.e. it's probably a state), it filters by that state, ignoring its location coordinates. 

This seems to work well for my use case, and I don't need to have a giant list of provinces in different languages and styles (like the PEI ones mentioned previously) since Google takes care of that for me.

We returned 0 in the query's distance field because I was ordering by distance in Twig and didn't know how else to get around it.

I didn't want to make a PR because it's pretty rough, but maybe someone else can use it:
https://github.com/missmatsuko/craft-smartmap/commit/aa67fff57b6581fdba2faeda45d7f90279d34614

**Comment by lindseydiloreto** [comment]
Fantastic, thanks @missmatsuko (and @tremby)! I'll take a look at those changes, and try to roll them back into the plugin. Nice work! 🍺

**Comment by lindseydiloreto** [comment]
Hey @missmatsuko, I'm _finally_ getting around to taking a closer look at your patch... but it looks like that link is now dead! 😕

Do you know where I could see a copy of those changes?

**Comment by missmatsuko** [comment]
Ah sorry, I wasn't using the repo so deleting it without remembering I'd linked to it here. I still had my local version so pushed it back up. Looks like the link works again now.

**Comment by lindseydiloreto** [comment]
Brilliant, thanks @missmatsuko! 👍

I'll take a look and see if/how it can be rolled back into the plugin.

**Comment by lindseydiloreto** [comment]
Alright @missmatsuko, I've added a new feature based on this concept. First, grab the latest version of the plugin from the `v3-dev` branch...

```
"doublesecretagency/craft-smartmap": "dev-v3-dev"
```

Then, check out this new page in the docs which explains how to use it...

https://www.doublesecretagency.com/plugins/smart-map/docs/using-a-filter-fallback-in-proximity-searches

Let me know if that does the trick for you!

**Comment by lindseydiloreto** [comment]
The "Filter Fallback" technique has been formally released in v3.3. 👍

If you bump into any issues regarding this new method, feel free to start a separate thread!
